ttl/ttlworker/job_manager.go

```diff
@@ -0,0 +1,571 @@
+
+	// the workers are shared between the loop goroutine and other sessions (e.g. manually resize workers through
+	// setting variables)
+	scanWorkers []worker
+	delWorkers  []worker
+
+	// infoSchemaCache and tableStatusCache are a cache stores the information from info schema and the tidb_ttl_table_status
+	// table. They don't need to be protected by mutex, because they are only used in job loop goroutine.
+	infoSchemaCache  *cache.InfoSchemaCache
+	tableStatusCache *cache.TableStatusCache
+
+	// runningJobs record all ttlJob waiting in local
+	// when a job for a table is created, it could spawn several scan tasks. If there are too many scan tasks, and they cannot
+	// be fully consumed by local scan workers, their states should be recorded in the runningJobs, so that we could continue
+	// to poll scan tasks from the job in the future when there are scan workers in idle.
+	runningJobs []*ttlJob
+
+	delCh         chan *ttlDeleteTask
+	notifyStateCh chan interface{}
+}
+
+// NewJobManager creates a new ttl job manager
+func NewJobManager(id string, sessPool sessionPool) (manager *JobManager) {
+	manager = &JobManager{}
+	manager.id = id
+	manager.sessPool = sessPool
+	manager.delCh = make(chan *ttlDeleteTask)
+
+	manager.init(manager.jobLoop)
+	manager.ctx = logutil.WithKeyValue(manager.ctx, "ttl-worker", "manager")
+
+	// TODO: read these variable from sysvars (from arguments, or other ways), rather than global variables in this pkg
+	manager.infoSchemaCache = cache.NewInfoSchemaCache(updateInfoSchemaCacheInterval)
+	manager.tableStatusCache = cache.NewTableStatusCache(updateTTLTableStatusCacheInterval)
+
+	manager.registerConfigFunctions()
+	return
+}
+
+func (m *JobManager) jobLoop() error {
+	se, err := getSession(m.sessPool)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = multierr.Combine(err, multierr.Combine(m.resizeScanWorkers(m.ctx, 0), m.resizeDelWorkers(m.ctx, 0)))
+		se.Close()
+	}()
+
+	scheduleTicker := time.Tick(jobManagerLoopTickerInterval)
+	updateHeartBeatTicker := time.Tick(jobManagerLoopTickerInterval)
+	jobFinishCheckTicker := time.Tick(jobManagerLoopTickerInterval)
+	updateScanTaskStateTicker := time.Tick(jobManagerLoopTickerInterval)
+	infoSchemaCacheUpdateTicker := time.Tick(m.infoSchemaCache.GetInterval())
+	tableStatusCacheUpdateTicker := time.Tick(m.tableStatusCache.GetInterval())
+	for {
+		now := se.Now()
+
+		select {
+		case <-m.ctx.Done():
+			return nil
+		case <-infoSchemaCacheUpdateTicker:
+			err := m.updateInfoSchemaCache(se)
+			if err != nil {
+				logutil.Logger(m.ctx).Warn("fail to update info schema cache", zap.Error(err))
+			}
+		case <-tableStatusCacheUpdateTicker:
+			err := m.updateTableStatusCache(se)
+			if err != nil {
+				logutil.Logger(m.ctx).Warn("fail to update table status cache", zap.Error(err))
+			}
+		case <-updateHeartBeatTicker:
+			updateHeartBeatCtx, cancel := context.WithTimeout(m.ctx, ttlInternalSQLTimeout)
+			err = m.updateHeartBeat(updateHeartBeatCtx, se)
```

Comment:
If heartbeat not success for a long time, will a job be discarded?

Reply:
It's a problem (but not too big). The `reschedule` logic checks whether the current owner is itself, so if the owner is updated by other nodes in the future, it'll not be scheduled.

The missing part is that it should also clean the job when the owner of this table entry is not itself :thinking: .

Reply:
Fixed.